croptalk/tools.py

```diff
@@ -158,10 +271,135 @@ def get_wfrp_commodities(reinsurance_yr: str, state_code: str, county_code: str)
+
+        },
+        {
+            "input": "What is the cost to grower under the APH policy for walnuts in Fresno county in California in 2023",
+            "query": "SELECT SUM(cost_to_grower * policy_sold_count)/SUM(cost_to_grower) AS average_cost_to_grower FROM sob_all_years WHERE commodity_year == 2023 AND county_name == 'Fresno' and state_abbreviation = 'CA' ORDER BY average_cost_to_grower DESC LIMIT 1"
+
+        },
+        {
+            "input": "What is the percentage of policies indemnified for Washakie county in Wyoming "
+                     "for Corn under the APH program",
+            "query": "SELECT SUM(policies_indemnified_count) * 100.0 / SUM(policies_sold_count) FROM sob_all_years WHERE county_name = 'Washakie' AND state_abbreviation = 'WY' AND insurance_pla_name_abbreviation == 'APH'"
+
+        },
+
+    ]
+
+
+system_prefix = """You are an agent designed to interact with a SQL database.
+Given an input question, create a syntactically correct {dialect} query to run, then look at the results of the query and return the answer.
+Unless the user specifies a specific number of examples they wish to obtain, always limit your query to at most {top_k} results.
+You can order the results by a relevant column to return the most interesting examples in the database.
+Never query for all the columns from a specific table, only ask for the relevant columns given the question.
+You have access to tools for interacting with the database.
+Only use the given tools. Only use the information returned by the tools to construct your final answer.
+You MUST double check your query before executing it. If you get an error while executing a query, rewrite the query and try again.
+
+Always use the sob_all_years table. 
+
+If a question is about an average, make sure to weight the metric by policy_sold_count.
+
+DO NOT make any DML statements (INSERT, UPDATE, DELETE, DROP etc.) to the database.
+
+If the question does not seem related to the database, just return "I don't know" as the answer.
+
+Here are some examples of user inputs and their corresponding SQL queries:"""
+
+examples = get_sob_sql_query_examples()
+
+example_selector = SemanticSimilarityExampleSelector.from_examples(
+    examples,
+    OpenAIEmbeddings(openai_api_key=os.environ["OPENAI_API_KEY"]),
+    FAISS,
+    k=5,
+    input_keys=["input"],
+)
+
+few_shot_prompt = FewShotPromptTemplate(
+    example_selector=example_selector,
+    example_prompt=PromptTemplate.from_template(
+        "User input: {input}\nSQL query: {query}"
+    ),
+    input_variables=["input", "dialect", "top_k"],
+    prefix=system_prefix,
+    suffix="",
+)
+
+full_prompt = ChatPromptTemplate.from_messages(
+    [
+        SystemMessagePromptTemplate(prompt=few_shot_prompt),
+        ("human", "{input}"),
+        MessagesPlaceholder("agent_scratchpad"),
+    ]
+)
+
+TOOLS = [get_sob_metrics_sql_agent, get_wfrp_commodities, get_sp_document]
```

Comment:
I noticed that validate_query is not in TOOLS... I think it's on purpose, but just wanted to validate with you...?

Reply:
yes, I integrated it directly with the sql agent